doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1208 @@
+  class attribute and the current ``dtype.type`` field considered
+  deprecated. This may be relaxed if a usecase arises.
+
+Additionally, existing methods (and C-side fields) will be provided, although
+fields such as the "kind" and and "char" may be set to invalid values on
+the C-side, and access may error on the Python side.
+
+Another example of methods that should be moved to the DType class are the
+various sorting functions, which shall be implemented by defining a method:
+
+* ``dtype_get_sort_function(self, sortkind="stable") -> sortfunction``
+
+Which must return ``NotImplemented`` if the given ``sortkind`` is not known.
+Similarly, any function implemented previously which cannot be removed will
+be implemented as a special method.
+Since these methods can be deprecated and new (renamed) replacements added,
+the API is not defined here and it is acceptable if it changes over time.
+
+For some of the current "methods" defined on the dtype, including sorting,
+a long term solution may be to instead create generalized ufuncs to provide
+the functionality.
+
+**Alternatives:**
+
+Some of these flags could be exposed to Python (or generally) by inheriting
+for example from a ``ParametricDType`` class. However, on the C-side as
+an implementation detail it seems simpler to provide a flag.
+This does not preclude the possibility of exposing a ``ParametricDType``
+to Python to represent the same thing.
+
+**Example:**
+
+The ``datetime64`` DType is considered parametric, due to its unit, and
+unlike a float64 has no canonical representation. The associated type
+is the ``np.datetime64`` scalar.
+
+
+**Issues and Details:**
+
+A potential DType such as ``Categorical`` will not be required to have a clear type
+associated with it. Instead the ``type`` may just be ``object``, its values are
+are arbitrary objects.  Unlike well-defined scalars, this ``type`` could
+not be used for the dtype discovery necessary for coercion
+(compare section `DType Discovery during Array Coercion`_).
+
+
+Coercion to and from Python Objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+**Motivation:**
+
+When storing a single value in an array or taking it out of the array
+a conversion must be made from the low-level representation to the Python
+object describing the scalar value.
+
+**Description:**
+
+Coercing to and from Python scalars requires two to three methods:
+
+1. ``__dtype_setitem__(self, item_pointer, value)``
+2. ``__dtype_getitem__(self, item_pointer, base_obj) -> object``
+   The ``base_obj`` should be ignored normally, it is provided *only* for
+   memory management purposes, pointing to an object owning the data.
+   It exists only to allow support of structured datatypes with subarrays
+   within NumPy, which (currently) return views into the array.
+   The function returns an equivalent Python scalar (i.e. typically a NumPy
+   scalar).
+3. ``__dtype_get_pyitem__(self, item_pointer, base_obj) -> object``
```

Comment:
Well, This is tricky and I may have to expand a bit. In my current branch, I already include casting (for scalars).  This means that `DType.__set_item__()` only needs to correctly be implemented for python types known to the DType (in particular, it must work for its own python scalar-type if that exists).
If the `DType` does not know what to do with the `value`, we will try to find the matching DType and then use casting.

This is currently not really done in NumPy (`float128` has a partial hack), and most types actually can handle everything in theory (since `value.__float__()` is correct for `float16`, `float32` and `float64`).

My current implementation is this:
```
if DType.type is type(value) and  DType.is_known_scalartype(value):
    # By default, should handle all basic python types
    return dtype.__dtype_setitem__(pointer, value)

cast_from_DType = disover_DType_from_pytype(type(value))
if cast_from_DType is None:
     # Not a scalar known to NumPy, so give original setitem a chance.
     return return dtype.__dtype_setitem__(pointer, value)

# We found a DType associated to the scalar type
cast_dtype = DType.discover_descriptor_from_pyobj(value)
tmp_buffer = buffer(size=descr.itemsize)
cast_dtype. __dtype_setitem__(buffer, value)
# cast tmp_buffer and write into pointer.
return
```
will add similar pseudo-code here as well.

Reply:
Probably good to expand as I'm not quite sure where the pseudo-code lives. Also not quite sure about the first `if` statement (why is being the right type not enough and does `value` needs to be checked) or the disappearance of pointer in the last clause. 

But I'll add that I've never even really looked properly at the current implementation, so if my requests for more clarification are becoming too much, do just ignore them!